// jinter/esm5/index.js
import { __awaiter, __generator } from "tslib";
import EventEmitter from 'eventemitter3';
var channel = 0;
var Servers = new Proxy({}, {
    get: function (obj, prop) {
        if (!obj[prop]) {
            obj[prop] = {
                ServerE: new EventEmitter(),
                ClientE: new EventEmitter(),
            };
        }
        return obj[prop];
    },
    deleteProperty: function (obj, prop) {
        obj[prop] = undefined;
        return true;
    },
});
var Client = /** @class */ (function () {
    function Client(id) {
        this.id = id;
        if (id) {
            var emt = Servers[id];
            this.ServerE = emt.ServerE;
            this.ClientE = emt.ClientE;
        }
    }
    Client.prototype.use = function (emitter) {
        this.emitter = emitter;
        return this;
    };
    Client.prototype.get = function (_a) {
        var server = _a.server, path = _a.path, data = _a.data;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.run(path, data, 'get', server)];
                    case 1: return [2 /*return*/, _b.sent()];
                }
            });
        });
    };
    /**
     * @TODO
     * return a success or error flag
     */
    Client.prototype.post = function (_a) {
        var server = _a.server, path = _a.path, body = _a.body;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.run(path, body, 'post', server)];
                    case 1: return [2 /*return*/, _b.sent()];
                }
            });
        });
    };
    Client.prototype.run = function (path, data, type, server) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, emitter, ServerE, ClientE;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this, emitter = _a.emitter, ServerE = _a.ServerE, ClientE = _a.ClientE;
                        if (emitter) {
                            this.emitter = void 0;
                            return [2 /*return*/, emitter(data)];
                        }
                        return [4 /*yield*/, new Promise(function (res) {
                                if (server) {
                                    var emt = Servers[server];
                                    ServerE = emt.ServerE;
                                    ClientE = emt.ClientE;
                                }
                                else if (_this.id) {
                                    server = _this.id;
                                }
                                else {
                                    // 没有server指向
                                    return res();
                                }
                                if (!ServerE || !ClientE)
                                    return res();
                                var serverEvName = type + "+" + server + "+" + path;
                                var hasServer = ServerE.eventNames().findIndex(function (v) { return v === serverEvName; }) > -1;
                                if (!hasServer)
                                    return res(void 0);
                                // 并发get
                                var thisChannel = channel++;
                                ClientE.once(type + "+client+" + server + "+" + path + ":" + thisChannel, function (resData) {
                                    res(resData);
                                });
                                ServerE.emit(serverEvName, { data: data, channel: thisChannel });
                            })];
                    case 1: return [2 /*return*/, _b.sent()];
                }
            });
        });
    };
    return Client;
}());
export { Client };
var Server = /** @class */ (function () {
    function Server(id) {
        this.id = id;
        this.events = [];
        var emt = Servers[id];
        this.ServerE = emt.ServerE;
        this.ClientE = emt.ClientE;
    }
    Server.prototype.prefixing = function (type, path) {
        return type + "+" + this.id + "+" + path;
    };
    Server.prototype.run = function (type, path, callback) {
        var _this = this;
        var _a = this, ServerE = _a.ServerE, ClientE = _a.ClientE;
        var serverName = this.prefixing(type, path);
        var eid = this.events.length;
        var clientName = type + "+client+" + this.id + "+" + path;
        var noServer = ServerE.eventNames().findIndex(function (v) { return v === serverName; }) === -1;
        if (!noServer && type === 'get') {
            throw new Error("[Jinter Error]: Get Serer(" + path + ") only can be listening once, you should remove one");
        }
        this.events.push({
            path: path,
            eid: eid,
            type: type,
            serverName: serverName,
            off: false,
            callback: callback,
        });
        ServerE &&
            ClientE &&
            noServer &&
            ServerE.on(serverName, function (_a) {
                var data = _a.data, channel = _a.channel;
                return __awaiter(_this, void 0, void 0, function () {
                    var events;
                    var _this = this;
                    return __generator(this, function (_b) {
                        events = this.events.filter(function (e) { return e.serverName === serverName && e.off === false; });
                        if (events.length > 0) {
                            events.forEach(function (e) { return __awaiter(_this, void 0, void 0, function () {
                                var res;
                                return __generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0:
                                            if (!e.callback) return [3 /*break*/, 2];
                                            return [4 /*yield*/, e.callback(data)];
                                        case 1:
                                            res = _a.sent();
                                            ClientE.emit(clientName + ':' + channel, res);
                                            return [3 /*break*/, 3];
                                        case 2:
                                            ClientE.emit(clientName + ':' + channel, void 0);
                                            _a.label = 3;
                                        case 3: return [2 /*return*/];
                                    }
                                });
                            }); });
                        }
                        else {
                            ClientE.emit(clientName + ':' + channel, void 0);
                        }
                        return [2 /*return*/];
                    });
                });
            });
        return eid;
    };
    Server.prototype.onGet = function (_a, callback) {
        var path = _a.path;
        return this.run('get', path, callback);
    };
    /**
     * @TODO
     * return a success or error flag
     */
    Server.prototype.onPost = function (_a, callback) {
        var path = _a.path;
        return this.run('post', path, callback);
    };
    Server.prototype.remove = function (type, path) {
        var name = this.prefixing(type, path);
        this.events = this.events.filter(function (v) { return !(v.path === path && v.type === type); });
        this.ServerE && this.ServerE.removeListener(name);
    };
    /**
     * 关闭某一个监听
     * @param eid 事件id
     */
    Server.prototype.off = function (eid, remove) {
        this.events = this.events.filter(function (e) {
            if (e.eid === eid) {
                e.off = true;
                return !remove;
            }
            else {
                return true;
            }
        });
        // type && path && this.remove(type, path)
    };
    /**
     * 开启某一个监听
     * @param eid 事件id
     */
    Server.prototype.on = function (eid) {
        this.events.forEach(function (e) {
            if (e.eid === eid) {
                e.off = false;
            }
        });
    };
    Server.prototype.close = function () {
        var _this = this;
        delete Servers[this.id];
        this.events.forEach(function (_a) {
            var path = _a.path;
            _this.remove('get', path);
            _this.remove('post', path);
        });
    };
    return Server;
}());
export { Server };
