// jinter/bundles/index.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('tslib')) :
	typeof define === 'function' && define.amd ? define(['exports', 'tslib'], factory) :
	(global = global || self, factory(global.Jinter = {}, global.tslib));
}(this, function (exports, tslib) { 'use strict';

	function createCommonjsModule(fn, module) {
		return module = { exports: {} }, fn(module, module.exports), module.exports;
	}

	var _eventemitter3_4_0_4_eventemitter3 = createCommonjsModule(function (module) {

	var has = Object.prototype.hasOwnProperty
	  , prefix = '~';

	/**
	 * Constructor to create a storage for our `EE` objects.
	 * An `Events` instance is a plain object whose properties are event names.
	 *
	 * @constructor
	 * @private
	 */
	function Events() {}

	//
	// We try to not inherit from `Object.prototype`. In some engines creating an
	// instance in this way is faster than calling `Object.create(null)` directly.
	// If `Object.create(null)` is not supported we prefix the event names with a
	// character to make sure that the built-in object properties are not
	// overridden or used as an attack vector.
	//
	if (Object.create) {
	  Events.prototype = Object.create(null);

	  //
	  // This hack is needed because the `__proto__` property is still inherited in
	  // some old browsers like Android 4, iPhone 5.1, Opera 11 and Safari 5.
	  //
	  if (!new Events().__proto__) prefix = false;
	}

	/**
	 * Representation of a single event listener.
	 *
	 * @param {Function} fn The listener function.
	 * @param {*} context The context to invoke the listener with.
	 * @param {Boolean} [once=false] Specify if the listener is a one-time listener.
	 * @constructor
	 * @private
	 */
	function EE(fn, context, once) {
	  this.fn = fn;
	  this.context = context;
	  this.once = once || false;
	}

	/**
	 * Add a listener for a given event.
	 *
	 * @param {EventEmitter} emitter Reference to the `EventEmitter` instance.
	 * @param {(String|Symbol)} event The event name.
	 * @param {Function} fn The listener function.
	 * @param {*} context The context to invoke the listener with.
	 * @param {Boolean} once Specify if the listener is a one-time listener.
	 * @returns {EventEmitter}
	 * @private
	 */
	function addListener(emitter, event, fn, context, once) {
	  if (typeof fn !== 'function') {
	    throw new TypeError('The listener must be a function');
	  }

	  var listener = new EE(fn, context || emitter, once)
	    , evt = prefix ? prefix + event : event;

	  if (!emitter._events[evt]) emitter._events[evt] = listener, emitter._eventsCount++;
	  else if (!emitter._events[evt].fn) emitter._events[evt].push(listener);
	  else emitter._events[evt] = [emitter._events[evt], listener];

	  return emitter;
	}

	/**
	 * Clear event by name.
	 *
	 * @param {EventEmitter} emitter Reference to the `EventEmitter` instance.
	 * @param {(String|Symbol)} evt The Event name.
	 * @private
	 */
	function clearEvent(emitter, evt) {
	  if (--emitter._eventsCount === 0) emitter._events = new Events();
	  else delete emitter._events[evt];
	}

	/**
	 * Minimal `EventEmitter` interface that is molded against the Node.js
	 * `EventEmitter` interface.
	 *
	 * @constructor
	 * @public
	 */
	function EventEmitter() {
	  this._events = new Events();
	  this._eventsCount = 0;
	}

	/**
	 * Return an array listing the events for which the emitter has registered
	 * listeners.
	 *
	 * @returns {Array}
	 * @public
	 */
	EventEmitter.prototype.eventNames = function eventNames() {
	  var names = []
	    , events
	    , name;

	  if (this._eventsCount === 0) return names;

	  for (name in (events = this._events)) {
	    if (has.call(events, name)) names.push(prefix ? name.slice(1) : name);
	  }

	  if (Object.getOwnPropertySymbols) {
	    return names.concat(Object.getOwnPropertySymbols(events));
	  }

	  return names;
	};

	/**
	 * Return the listeners registered for a given event.
	 *
	 * @param {(String|Symbol)} event The event name.
	 * @returns {Array} The registered listeners.
	 * @public
	 */
	EventEmitter.prototype.listeners = function listeners(event) {
	  var evt = prefix ? prefix + event : event
	    , handlers = this._events[evt];

	  if (!handlers) return [];
	  if (handlers.fn) return [handlers.fn];

	  for (var i = 0, l = handlers.length, ee = new Array(l); i < l; i++) {
	    ee[i] = handlers[i].fn;
	  }

	  return ee;
	};

	/**
	 * Return the number of listeners listening to a given event.
	 *
	 * @param {(String|Symbol)} event The event name.
	 * @returns {Number} The number of listeners.
	 * @public
	 */
	EventEmitter.prototype.listenerCount = function listenerCount(event) {
	  var evt = prefix ? prefix + event : event
	    , listeners = this._events[evt];

	  if (!listeners) return 0;
	  if (listeners.fn) return 1;
	  return listeners.length;
	};

	/**
	 * Calls each of the listeners registered for a given event.
	 *
	 * @param {(String|Symbol)} event The event name.
	 * @returns {Boolean} `true` if the event had listeners, else `false`.
	 * @public
	 */
	EventEmitter.prototype.emit = function emit(event, a1, a2, a3, a4, a5) {
	  var evt = prefix ? prefix + event : event;

	  if (!this._events[evt]) return false;

	  var listeners = this._events[evt]
	    , len = arguments.length
	    , args
	    , i;

	  if (listeners.fn) {
	    if (listeners.once) this.removeListener(event, listeners.fn, undefined, true);

	    switch (len) {
	      case 1: return listeners.fn.call(listeners.context), true;
	      case 2: return listeners.fn.call(listeners.context, a1), true;
	      case 3: return listeners.fn.call(listeners.context, a1, a2), true;
	      case 4: return listeners.fn.call(listeners.context, a1, a2, a3), true;
	      case 5: return listeners.fn.call(listeners.context, a1, a2, a3, a4), true;
	      case 6: return listeners.fn.call(listeners.context, a1, a2, a3, a4, a5), true;
	    }

	    for (i = 1, args = new Array(len -1); i < len; i++) {
	      args[i - 1] = arguments[i];
	    }

	    listeners.fn.apply(listeners.context, args);
	  } else {
	    var length = listeners.length
	      , j;

	    for (i = 0; i < length; i++) {
	      if (listeners[i].once) this.removeListener(event, listeners[i].fn, undefined, true);

	      switch (len) {
	        case 1: listeners[i].fn.call(listeners[i].context); break;
	        case 2: listeners[i].fn.call(listeners[i].context, a1); break;
	        case 3: listeners[i].fn.call(listeners[i].context, a1, a2); break;
	        case 4: listeners[i].fn.call(listeners[i].context, a1, a2, a3); break;
	        default:
	          if (!args) for (j = 1, args = new Array(len -1); j < len; j++) {
	            args[j - 1] = arguments[j];
	          }

	          listeners[i].fn.apply(listeners[i].context, args);
	      }
	    }
	  }

	  return true;
	};

	/**
	 * Add a listener for a given event.
	 *
	 * @param {(String|Symbol)} event The event name.
	 * @param {Function} fn The listener function.
	 * @param {*} [context=this] The context to invoke the listener with.
	 * @returns {EventEmitter} `this`.
	 * @public
	 */
	EventEmitter.prototype.on = function on(event, fn, context) {
	  return addListener(this, event, fn, context, false);
	};

	/**
	 * Add a one-time listener for a given event.
	 *
	 * @param {(String|Symbol)} event The event name.
	 * @param {Function} fn The listener function.
	 * @param {*} [context=this] The context to invoke the listener with.
	 * @returns {EventEmitter} `this`.
	 * @public
	 */
	EventEmitter.prototype.once = function once(event, fn, context) {
	  return addListener(this, event, fn, context, true);
	};

	/**
	 * Remove the listeners of a given event.
	 *
	 * @param {(String|Symbol)} event The event name.
	 * @param {Function} fn Only remove the listeners that match this function.
	 * @param {*} context Only remove the listeners that have this context.
	 * @param {Boolean} once Only remove one-time listeners.
	 * @returns {EventEmitter} `this`.
	 * @public
	 */
	EventEmitter.prototype.removeListener = function removeListener(event, fn, context, once) {
	  var evt = prefix ? prefix + event : event;

	  if (!this._events[evt]) return this;
	  if (!fn) {
	    clearEvent(this, evt);
	    return this;
	  }

	  var listeners = this._events[evt];

	  if (listeners.fn) {
	    if (
	      listeners.fn === fn &&
	      (!once || listeners.once) &&
	      (!context || listeners.context === context)
	    ) {
	      clearEvent(this, evt);
	    }
	  } else {
	    for (var i = 0, events = [], length = listeners.length; i < length; i++) {
	      if (
	        listeners[i].fn !== fn ||
	        (once && !listeners[i].once) ||
	        (context && listeners[i].context !== context)
	      ) {
	        events.push(listeners[i]);
	      }
	    }

	    //
	    // Reset the array, or remove it completely if we have no more listeners.
	    //
	    if (events.length) this._events[evt] = events.length === 1 ? events[0] : events;
	    else clearEvent(this, evt);
	  }

	  return this;
	};

	/**
	 * Remove all listeners, or those of the specified event.
	 *
	 * @param {(String|Symbol)} [event] The event name.
	 * @returns {EventEmitter} `this`.
	 * @public
	 */
	EventEmitter.prototype.removeAllListeners = function removeAllListeners(event) {
	  var evt;

	  if (event) {
	    evt = prefix ? prefix + event : event;
	    if (this._events[evt]) clearEvent(this, evt);
	  } else {
	    this._events = new Events();
	    this._eventsCount = 0;
	  }

	  return this;
	};

	//
	// Alias methods names because people roll like that.
	//
	EventEmitter.prototype.off = EventEmitter.prototype.removeListener;
	EventEmitter.prototype.addListener = EventEmitter.prototype.on;

	//
	// Expose the prefix.
	//
	EventEmitter.prefixed = prefix;

	//
	// Allow `EventEmitter` to be imported as module namespace.
	//
	EventEmitter.EventEmitter = EventEmitter;

	//
	// Expose the module.
	//
	{
	  module.exports = EventEmitter;
	}
	});

	var channel = 0;
	var Servers = new Proxy({}, {
	    get: function (obj, prop) {
	        if (!obj[prop]) {
	            obj[prop] = {
	                ServerE: new _eventemitter3_4_0_4_eventemitter3(),
	                ClientE: new _eventemitter3_4_0_4_eventemitter3(),
	            };
	        }
	        return obj[prop];
	    },
	    deleteProperty: function (obj, prop) {
	        obj[prop] = undefined;
	        return true;
	    },
	});
	var Client = /** @class */ (function () {
	    function Client(id) {
	        this.id = id;
	        if (id) {
	            var emt = Servers[id];
	            this.ServerE = emt.ServerE;
	            this.ClientE = emt.ClientE;
	        }
	    }
	    Client.prototype.use = function (emitter) {
	        this.emitter = emitter;
	        return this;
	    };
	    Client.prototype.get = function (_a) {
	        var server = _a.server, path = _a.path, data = _a.data;
	        return tslib.__awaiter(this, void 0, void 0, function () {
	            return tslib.__generator(this, function (_b) {
	                switch (_b.label) {
	                    case 0: return [4 /*yield*/, this.run(path, data, 'get', server)];
	                    case 1: return [2 /*return*/, _b.sent()];
	                }
	            });
	        });
	    };
	    /**
	     * @TODO
	     * return a success or error flag
	     */
	    Client.prototype.post = function (_a) {
	        var server = _a.server, path = _a.path, body = _a.body;
	        return tslib.__awaiter(this, void 0, void 0, function () {
	            return tslib.__generator(this, function (_b) {
	                switch (_b.label) {
	                    case 0: return [4 /*yield*/, this.run(path, body, 'post', server)];
	                    case 1: return [2 /*return*/, _b.sent()];
	                }
	            });
	        });
	    };
	    Client.prototype.run = function (path, data, type, server) {
	        return tslib.__awaiter(this, void 0, void 0, function () {
	            var _a, emitter, ServerE, ClientE;
	            var _this = this;
	            return tslib.__generator(this, function (_b) {
	                switch (_b.label) {
	                    case 0:
	                        _a = this, emitter = _a.emitter, ServerE = _a.ServerE, ClientE = _a.ClientE;
	                        if (emitter) {
	                            this.emitter = void 0;
	                            return [2 /*return*/, emitter(data)];
	                        }
	                        return [4 /*yield*/, new Promise(function (res) {
	                                if (server) {
	                                    var emt = Servers[server];
	                                    ServerE = emt.ServerE;
	                                    ClientE = emt.ClientE;
	                                }
	                                else if (_this.id) {
	                                    server = _this.id;
	                                }
	                                else {
	                                    // 没有server指向
	                                    return res();
	                                }
	                                if (!ServerE || !ClientE)
	                                    return res();
	                                var serverEvName = type + "+" + server + "+" + path;
	                                var hasServer = ServerE.eventNames().findIndex(function (v) { return v === serverEvName; }) > -1;
	                                if (!hasServer)
	                                    return res(void 0);
	                                // 并发get
	                                var thisChannel = channel++;
	                                ClientE.once(type + "+client+" + server + "+" + path + ":" + thisChannel, function (resData) {
	                                    res(resData);
	                                });
	                                ServerE.emit(serverEvName, { data: data, channel: thisChannel });
	                            })];
	                    case 1: return [2 /*return*/, _b.sent()];
	                }
	            });
	        });
	    };
	    return Client;
	}());
	var Server = /** @class */ (function () {
	    function Server(id) {
	        this.id = id;
	        this.events = [];
	        var emt = Servers[id];
	        this.ServerE = emt.ServerE;
	        this.ClientE = emt.ClientE;
	    }
	    Server.prototype.prefixing = function (type, path) {
	        return type + "+" + this.id + "+" + path;
	    };
	    Server.prototype.run = function (type, path, callback) {
	        var _this = this;
	        var _a = this, ServerE = _a.ServerE, ClientE = _a.ClientE;
	        var serverName = this.prefixing(type, path);
	        var eid = this.events.length;
	        var clientName = type + "+client+" + this.id + "+" + path;
	        var noServer = ServerE.eventNames().findIndex(function (v) { return v === serverName; }) === -1;
	        if (!noServer && type === 'get') {
	            throw new Error("[Jinter Error]: Get Serer(" + path + ") only can be listening once, you should remove one");
	        }
	        this.events.push({
	            path: path,
	            eid: eid,
	            type: type,
	            serverName: serverName,
	            off: false,
	            callback: callback,
	        });
	        ServerE &&
	            ClientE &&
	            noServer &&
	            ServerE.on(serverName, function (_a) {
	                var data = _a.data, channel = _a.channel;
	                return tslib.__awaiter(_this, void 0, void 0, function () {
	                    var events;
	                    var _this = this;
	                    return tslib.__generator(this, function (_b) {
	                        events = this.events.filter(function (e) { return e.serverName === serverName && e.off === false; });
	                        if (events.length > 0) {
	                            events.forEach(function (e) { return tslib.__awaiter(_this, void 0, void 0, function () {
	                                var res;
	                                return tslib.__generator(this, function (_a) {
	                                    switch (_a.label) {
	                                        case 0:
	                                            if (!e.callback) return [3 /*break*/, 2];
	                                            return [4 /*yield*/, e.callback(data)];
	                                        case 1:
	                                            res = _a.sent();
	                                            ClientE.emit(clientName + ':' + channel, res);
	                                            return [3 /*break*/, 3];
	                                        case 2:
	                                            ClientE.emit(clientName + ':' + channel, void 0);
	                                            _a.label = 3;
	                                        case 3: return [2 /*return*/];
	                                    }
	                                });
	                            }); });
	                        }
	                        else {
	                            ClientE.emit(clientName + ':' + channel, void 0);
	                        }
	                        return [2 /*return*/];
	                    });
	                });
	            });
	        return eid;
	    };
	    Server.prototype.onGet = function (_a, callback) {
	        var path = _a.path;
	        return this.run('get', path, callback);
	    };
	    /**
	     * @TODO
	     * return a success or error flag
	     */
	    Server.prototype.onPost = function (_a, callback) {
	        var path = _a.path;
	        return this.run('post', path, callback);
	    };
	    Server.prototype.remove = function (type, path) {
	        var name = this.prefixing(type, path);
	        this.events = this.events.filter(function (v) { return !(v.path === path && v.type === type); });
	        this.ServerE && this.ServerE.removeListener(name);
	    };
	    /**
	     * 关闭某一个监听
	     * @param eid 事件id
	     */
	    Server.prototype.off = function (eid, remove) {
	        this.events = this.events.filter(function (e) {
	            if (e.eid === eid) {
	                e.off = true;
	                return !remove;
	            }
	            else {
	                return true;
	            }
	        });
	        // type && path && this.remove(type, path)
	    };
	    /**
	     * 开启某一个监听
	     * @param eid 事件id
	     */
	    Server.prototype.on = function (eid) {
	        this.events.forEach(function (e) {
	            if (e.eid === eid) {
	                e.off = false;
	            }
	        });
	    };
	    Server.prototype.close = function () {
	        var _this = this;
	        delete Servers[this.id];
	        this.events.forEach(function (_a) {
	            var path = _a.path;
	            _this.remove('get', path);
	            _this.remove('post', path);
	        });
	    };
	    return Server;
	}());

	exports.Client = Client;
	exports.Server = Server;

	Object.defineProperty(exports, '__esModule', { value: true });

}));
